### ISSUE TITLE
Structured concurrency for gevent

### ISSUE THREAD

**Issue by lewoudar** [gevent/gevent#1697]
Hi,
this is not an issue but a proposal for gevent improvement. I would like to talk about *structured concurrency* which I think is a promising concept for the future of concurrent programming. In short, it guarantees that tasks spawned in a block of code are finished in one way or another. Initial use of this term is attributed to Martin Sústrik in this [blog post](https://250bpm.com/blog:71/).

I discovered this concept through the [trio](https://trio.readthedocs.io/en/stable/) library, a beautfiful asynchronous concurrency library leveraging the `async/await` syntax. The creator Nathanael J Smith (who is also a python core developer) uses the principle of [nurseries](https://trio.readthedocs.io/en/stable/tutorial.html#okay-let-s-see-something-cool-already) to guarantee the lifetime of tasks. In gevent it could be something like this

```python
import gevent

with gevent.create_task_group() as tg:
  tg.spawn(task1, arg1, ..)
  tg.spawn(task2, arg1, ..)
```

One can argue that the [gevent.Pool](http://www.gevent.org/api/gevent.pool.html) does the same thing, but I dont think so, since tasks spawned by the pool can spawed other tasks whose lifespan is independent of the pool's lifespan. In trio, only nurseries can spawn tasks to ensure total control over the tasks lifespan.

In my recent project [pyscalpel](https://scalpel.readthedocs.io/en/latest/), I have to simulate nurseries by using a pool combined with a `JoinableQueue` :smile: 
I know that it will be a big shift gevent because it probably means to redesign many concepts of the library but I think it's worth it. For me it helps reduce bugs and unpredictable behaviours in a program.

If you want more readings to mature your knowledge (since I'm not an expert on this subject), here are a few links:
- [Notes on structured concurrency, or go statement considered harmful](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/)
- The trio [design](https://trio.readthedocs.io/en/stable/design.html)
- A [wiki](https://trio.discourse.group/t/structured-concurrency-resources/21) with other resources related to structured concurrency

Let me know what you think about it :blush:

**Comment by navytux** [comment]
You can already have it for gevent via [pygolang](https://pypi.org/project/pygolang) and [`sync.WorkGroup`](https://pypi.org/project/pygolang/#concurrency):

```py
from golang import defer
from golang import sync

    ctx = ...
    wg = sync.WorkGroup(ctx)
    defer(wg.wait)
    wg.go(task1, arg1, ...)
    wg.go(task2, arg2, ...)
```

because pygolang provides both "threads" and "gevent" runtimes, and so the same program can be run with `go` spawning either OS-level threads or gevent greenlets depending on how the program was run.

For "with" syntax

```py
from golang import sync

    ctx = ...
    with sync.WorkGroup(ctx) as wg:
        wg.go(task1, arg1, ...)
        wg.go(task2, arg2, ...)
```

please see https://lab.nexedi.com/nexedi/pygolang/merge_requests/12.

**Comment by lewoudar** [comment]
Aw @navytux this is interesting! To be honest I'm not a huge fan of the context object that we have to pass everywhere, but I will keep one eye on your project. It is great to have "some resistants" of the async/await trend. Seems like my issue here inspired you to add this construct in your project, it's cool 😄 
Nevertherless, it would be great that gevent introduces this construct in future releases

**Comment by jamadden** [comment]
I don't think this is something gevent should or will do. gevent provides the primitives and other projects like pygolang can build on top of that as they see fit.

**Comment by lewoudar** [comment]
No offense, but I think you should reconsider your position. Add a `TaskGroup` would not be too different from the `gevent.Pool` which is also a sort of high level construct. It will help many "new" developers to write correct concurrent code IMHO.

**Comment by navytux** [comment]
@lewoudar, thanks for feedback. `sync.WorkGroup` was there from [long ago](https://lab.nexedi.com/nexedi/pygolang/commit/9ee7ba91) because I use it in wendelin.core (see e.g. [1](https://lab.nexedi.com/kirr/wendelin.core/blob/ac6e120f/wcfs/__init__.py#L230-284), [2](https://lab.nexedi.com/kirr/wendelin.core/blob/ac6e120f/wcfs/wcfs_test.py#L958-1008), [3](https://lab.nexedi.com/kirr/wendelin.core/blob/ac6e120f/wcfs/wcfs_test.py#L1276-1297), [4](https://lab.nexedi.com/kirr/wendelin.core/blob/ac6e120f/wcfs/wcfs_test.py#L1389-1409), [5](https://lab.nexedi.com/kirr/wendelin.core/blob/ac6e120f/wcfs/client/wcfs.cpp#L327-332)). Having "with" support for it was long on my mind, and I finally decided to implement it after seeing your comment yesterday. So thanks indeed :)

Context carries cancellation signal and work-scoped values. I understand there are pros and cons for being explicit instead of implicitly carrying that information in thread/greenlet-local storage. But even in Python world they [say](https://www.python.org/dev/peps/pep-0020/) that "explicit is better than implicit", isn't it? (see also `import this`).

Thanks again for feedback,  
Kirill

**Comment by lewoudar** [comment]
@navytux you are welcome and yeah, explicit is better than implicit. I don't have enough background on the context object to have a definite opinion on it.

**Comment by avanov** [comment]
@lewoudar why wouldn't trio implement it in its codebase with gevent as an alternative coroutine backend?  I don't see why it's preferred for a coroutine library to implement and maintain high level interfaces and other building blocks as part of their codebase. The pool can be moved into a separate library that supports multiple backends as well.

**Comment by lewoudar** [comment]
Hi @avanov, it would be good if you read this [article](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/] of trio author. The nursery object is not a *high level interface* it is the **only way** to make concurrency with trio.

**Comment by avanov** [comment]
I've read it, still the question is the same, why the Nursery class cannot be baked off gevent's primitives? [This](https://github.com/python-trio/trio/blob/master/trio/_core/_io_epoll.py#L188), [this](https://github.com/python-trio/trio/blob/32302ebfde0a8bd62a941945c178e1583f6e2811/trio/_core/_run.py#L1414), [this](https://github.com/python-trio/trio/blob/32302ebfde0a8bd62a941945c178e1583f6e2811/trio/_core/_run.py#L1940), and [this](https://github.com/python-trio/trio/blob/32302ebfde0a8bd62a941945c178e1583f6e2811/trio/_core/_run.py#L1059) suggest that what Trio is doing in a nutshell is synchronising coroutines through a queue in particular selected strategy that they define as "Structured Concurrency". Everything else on top of that is a high level interface. Their `TheIOManger` uses epoll on linux, gevent uses epoll as well through libev/libuv. Their `Task` is a wrapper for a lower level primitive that in gevent is represented through `Greenlet`.  I don't see any particular obstacle for their `Task`s to be gevent's `Greenlet`s, and a queue to be gevent's `Queue` as one of the possible implementation of lower-level event loop primitives.

**Comment by lewoudar** [comment]
> still the question is the same, why the Nursery class cannot be baked off gevent's primitives? 

Hum.. may be I don't understand, but I'm advocating to a first-class support of Nursery in gevent because I think it can help many users to write correct concurrent code. Beginner programmers will not accidentally leave tasks running in the background without making sure that they don't end up somehow. Also they will be forced to take into account errors that are always propagated and not *stifled* as in the case of gevent. Of course in gevent, you can always check if errors happened in a a greenlet using the `get` method, but I am sure that many *forget* to do that.

Nurseries are not just a "high-level construct" to spawn tasks, they imply many good programming practices IMHO :)

**Comment by navytux** [comment]
For the reference: `with` support for `sync.WorkGroup` was released as part of pygolang [0.0.9](https://pypi.org/project/pygolang/#pygolang-change-history).
